File: web/src/lib/crypto.ts
import type Sodium from "libsodium-wrappers";

// libsodium is ~450KB of wasm/asm and is the single biggest thing the app can
// download. It's loaded lazily (and cached as one promise) instead of statically
// imported so it lands in its own chunk: the entry bundle no longer carries it,
// and a business-logic release doesn't invalidate it in the browser cache.
//
// Every synchronous export below reaches for `lib` — all of them are already
// documented as "requires ready()", and every call site awaits `ready()` during
// startup, so the throw is a contract violation, not a reachable state.
let lib: typeof Sodium | undefined;
let loading: Promise<typeof Sodium> | undefined;

export async function ready(): Promise<void> {
  loading ??= import("libsodium-wrappers").then(async (m) => {
    const s = m.default;
    await s.ready;
    lib = s;
    return s;
  });
  try {
    await loading;
  } catch (e) {
    loading = undefined; // a failed fetch (flaky network) must not poison every retry
    throw e;
  }
}

function sodiumSync(): typeof Sodium {
  if (!lib) throw new Error("libsodium not initialised — await ready() first");
  return lib;
}

// Buffers that cross the Web Crypto / DOM boundary must be explicitly
// ArrayBuffer-backed: TS's generic `Uint8Array<ArrayBufferLike>` is rejected by
// `importKey`/`encrypt`/`decrypt`, which demand `Uint8Array<ArrayBuffer>`. Every
// buffer here is ArrayBuffer-backed at runtime (never SharedArrayBuffer).
type Bytes = Uint8Array<ArrayBuffer>;

export interface KeyPair {
  publicKey: Uint8Array;
  privateKey: Uint8Array;
}

export interface SessionKeys {
  send: CryptoKey;
  recv: CryptoKey;
  /** HMAC key both sides derive to the SAME value, used to authenticate resume
   *  signalling (see signResume). Separate from send/recv on purpose: it signs
   *  attacker-visible plaintext, so it must not be the key protecting content. */
  resumeAuth: CryptoKey;
  /** AEAD keys for the message stream, domain-separated from send/recv.
   *
   *  送文件那条流的 nonce 安全性建立在"只有一个生产者推进 seq"之上（见 transfer.ts
   *  的 Sender.seq），而消息是**第二个**生产者：它由 UI 事件触发，和一个在 yield 处
   *  挂起、可能永远不恢复的异步生成器交错。两个生产者共用一个密钥正是 AES-GCM 复用
   *  nonce 的那个形状。派生出独立子密钥之后，消息流有自己从 0 开始的计数器，交错这个
   *  问题就消失了，而不是被回答了。 */
  textSend: CryptoKey;
  textRecv: CryptoKey;
}

export function generateKeyPair(): KeyPair {
  const kp = sodiumSync().crypto_kx_keypair();
  return { publicKey: kp.publicKey, privateKey: kp.privateKey };
}

async function importAesKey(raw: Bytes): Promise<CryptoKey> {
  return crypto.subtle.importKey("raw", raw, "AES-GCM", false, [
    "encrypt",
    "decrypt",
  ]);
}

export async function deriveSession(
  role: "initiator" | "responder",
  self: KeyPair,
  peerPublic: Uint8Array,
): Promise<SessionKeys> {
  // crypto_kx gives a (rx, tx) pair; client/server roles produce mirror-image
  // keys so that one side's tx equals the other side's rx.
  const keys =
    role === "initiator"
      ? sodiumSync().crypto_kx_client_session_keys(
          self.publicKey,
          self.privateKey,
          peerPublic,
        )
      : sodiumSync().crypto_kx_server_session_keys(
          self.publicKey,
          self.privateKey,
          peerPublic,
        );
  // libsodium-wrappers types its outputs as the bare generic `Uint8Array`, but
  // its session keys are always ArrayBuffer-backed at runtime.
  return {
    send: await importAesKey(keys.sharedTx as Bytes),
    recv: await importAesKey(keys.sharedRx as Bytes),
    resumeAuth: await deriveResumeAuth(keys.sharedTx as Bytes, keys.sharedRx as Bytes),
    textSend: await importAesKey(textKeyBytes(keys.sharedTx as Bytes) as Bytes),
    textRecv: await importAesKey(textKeyBytes(keys.sharedRx as Bytes) as Bytes),
  };
}

/** Domain separation: this key must never produce a tag that could be replayed
 *  into any other context that might later hash the same session secrets. */
const RESUME_AUTH_DOMAIN = "relayium-resume-auth-v1\0";

/** Derive the shared resume-authentication key from the session secrets.
 *
 *  crypto_kx hands the two sides MIRRORED secrets (one's tx is the other's rx),
 *  so hashing them in local order would give the peers two different keys. They
 *  are sorted first — the pair as a SET is identical on both sides, so the sort
 *  is what makes this derivation symmetric without any extra round trip. */
async function deriveResumeAuth(tx: Bytes, rx: Bytes): Promise<CryptoKey> {
  const [a, b] = compareBytes(tx, rx) <= 0 ? [tx, rx] : [rx, tx];
  const domain = new TextEncoder().encode(RESUME_AUTH_DOMAIN);
  const input = new Uint8Array(domain.length + a.length + b.length);
  input.set(domain, 0);
  input.set(a, domain.length);
  input.set(b, domain.length + a.length);
  // `null` key = unkeyed BLAKE2b, byte-identical to omitting the argument; the
  // resolved libsodium-wrappers types mark it required, so pass it explicitly the
  // way commitKey/sas/textKeyBytes already do. Pinned by a vector test: the derived
  // key still reproduces crypto-vectors.json's resumeAuth.mac, which the Swift port
  // pins independently.
  const raw = sodiumSync().crypto_generichash(32, input, null) as Bytes;
  return crypto.subtle.importKey("raw", raw, { name: "HMAC", hash: "SHA-256" }, false, [
    "sign",
    "verify",
  ]);
}

/** Domain separation for the message stream's keys. Changing this string is a
 *  wire break: regenerate apps/RelayiumKit/Tests/Fixtures/crypto-vectors.json and
 *  move web + Swift together (docs/protocol/relayium-crypto-v1.md). */
export const TEXT_KEY_DOMAIN = "relayium-text-v1\0";

/**
 * The 32-byte AEAD key for ONE direction of the message stream.
 *
 * Unlike deriveResumeAuth this deliberately does **not** sort its inputs, and
 * must not. That key is shared, so it has to be symmetric; these are per
 * direction. crypto_kx already hands the peers mirrored secrets — one side's tx
 * is the other's rx — so hashing each one locally lines the directions up with
 * no extra round trip. Sorting would collapse both directions onto a single key
 * and put two producers back on one nonce counter, which is the exact hazard the
 * separate key exists to remove.
 *
 * The domain prefix is what stops this being a bare hash of a session secret,
 * replayable into any other context that might later hash the same secret.
 */
export function textKeyBytes(sessionKey: Uint8Array): Uint8Array {
  const domain = new TextEncoder().encode(TEXT_KEY_DOMAIN);
  const input = new Uint8Array(domain.length + sessionKey.length);
  input.set(domain, 0);
  input.set(sessionKey, domain.length);
  // `null` key = unkeyed BLAKE2b, byte-identical to omitting the argument; the
  // resolved libsodium-wrappers types mark it required, so pass it (same as
  // commitKey/sas/deriveResumeAuth).
  return sodiumSync().crypto_generichash(32, input, null) as Bytes;
}

function compareBytes(x: Uint8Array, y: Uint8Array): number {
  for (let i = 0; i < Math.min(x.length, y.length); i++) {
    if (x[i] !== y[i]) return x[i] - y[i];
  }
  return x.length - y.length;
}

/** Tag for a resume signalling message, base64. */
export async function signResume(key: CryptoKey, payload: string): Promise<string> {
  const mac = await crypto.subtle.sign("HMAC", key, new TextEncoder().encode(payload));
  return btoa(String.fromCharCode(...new Uint8Array(mac)));
}

/** Check a resume tag. A malformed/absent tag is a failure, never a pass —
 *  falling back to "unauthenticated is fine" would make the whole binding
 *  optional at the attacker's choosing. */
export async function verifyResume(
  key: CryptoKey,
  payload: string,
  mac: string | undefined,
): Promise<boolean> {
  if (!mac) return false;
  // Explicitly ArrayBuffer-backed: Web Crypto's BufferSource rejects the generic
  // `Uint8Array<ArrayBufferLike>` that Uint8Array.from infers, and every buffer in
  // this file is ArrayBuffer-backed at runtime (never SharedArrayBuffer).
  let sig: Bytes;
  try {
    sig = Uint8Array.from(atob(mac), (c) => c.charCodeAt(0)) as Bytes;
  } catch {
    return false;
  }
  return crypto.subtle.verify("HMAC", key, sig, new TextEncoder().encode(payload));
}

function nonceFromSeq(seq: number): Bytes {
  const n = new Uint8Array(12);
  const view = new DataView(n.buffer);
  // high 4 bytes zero; low 8 bytes hold the counter (supports >2^53 frames anyway).
  view.setUint32(4, Math.floor(seq / 2 ** 32));
  view.setUint32(8, seq >>> 0);
  return n;
}

export async function seal(
  key: CryptoKey,
  seq: number,
  plaintext: Bytes,
): Promise<Bytes> {
  const ct = await crypto.subtle.encrypt(
    { name: "AES-GCM", iv: nonceFromSeq(seq) },
    key,
    plaintext,
  );
  return new Uint8Array(ct);
}

export async function open(
  key: CryptoKey,
  seq: number,
  ciphertext: Bytes,
): Promise<Bytes> {
  const pt = await crypto.subtle.decrypt(
    { name: "AES-GCM", iv: nonceFromSeq(seq) },
    key,
    ciphertext,
  );
  return new Uint8Array(pt);
}

// ── SAS commitment (commit-then-reveal handshake) ────────────────────────────
// A 6-digit SAS is only ~20 bits, so a malicious signaling relay that sees both
// real public keys before choosing its own could brute-force a colliding code
// and MITM the session. To stop that we bind each side's public key with a
// commitment `C = BLAKE2b(pub || nonce)` that is exchanged *before* either side
// reveals its real public key. An attacker must therefore commit to its keys
// without having seen the peer's key, removing the post-hoc collision freedom.
export const COMMIT_BYTES = 32;
export const NONCE_BYTES = 32;

/** Fresh 32-byte commitment nonce. Requires ready(). */
export function randomNonce(): Uint8Array {
  return sodiumSync().randombytes_buf(NONCE_BYTES);
}

/** Commitment to a public key: 32-byte BLAKE2b(pub || nonce). */
export function commitKey(pub: Uint8Array, nonce: Uint8Array): Uint8Array {
  const combined = new Uint8Array(pub.length + nonce.length);
  combined.set(pub, 0);
  combined.set(nonce, pub.length);
  return sodiumSync().crypto_generichash(COMMIT_BYTES, combined, null);
}

/** Constant-time check that `commit` opens to (pub, nonce). False on any
 *  mismatch, including a malformed/short commitment. */
export function verifyCommit(
  commit: Uint8Array,
  pub: Uint8Array,
  nonce: Uint8Array,
): boolean {
  const expected = commitKey(pub, nonce);
  if (commit.length !== expected.length) return false;
  return sodiumSync().memcmp(commit as Bytes, expected as Bytes);
}

export function sas(self: Uint8Array, peer: Uint8Array): string {
  // Order-independent: sort the two public keys before hashing.
  const [a, b] = compare(self, peer) <= 0 ? [self, peer] : [peer, self];
  const combined = new Uint8Array(a.length + b.length);
  combined.set(a, 0);
  combined.set(b, a.length);
  // `null` key = unkeyed BLAKE2b (byte-identical to omitting it); the resolved
  // libsodium-wrappers types mark the key parameter as required, so pass it.
  const digest = sodiumSync().crypto_generichash(8, combined, null);
  const view = new DataView(digest.buffer, digest.byteOffset, digest.byteLength);
  const num = (view.getUint32(0) ^ view.getUint32(4)) >>> 0;
  return (num % 1_000_000).toString().padStart(6, "0");
}

function compare(a: Uint8Array, b: Uint8Array): number {
  const len = Math.min(a.length, b.length);
  for (let i = 0; i < len; i++) {
    if (a[i] !== b[i]) return a[i] - b[i];
  }
  return a.length - b.length;
}

File: apps/android/protocol/src/main/kotlin/com/relayium/protocol/Crypto.kt
package com.relayium.protocol

import java.security.GeneralSecurityException
import java.security.MessageDigest
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec
import org.bouncycastle.crypto.digests.Blake2bDigest
import org.bouncycastle.math.ec.rfc7748.X25519

/**
 * The crypto layer of `link/1`, byte-pinned to `docs/protocol/relayium-crypto-v1.md`
 * and to `apps/RelayiumKit/Tests/Fixtures/crypto-vectors.json`.
 *
 * ## What is borrowed and what is not
 *
 * Two primitives come from BouncyCastle because the JDK and Android have
 * neither: raw X25519 scalar multiplication (`rfc7748.X25519`) and BLAKE2b
 * (`Blake2bDigest`). Everything else — AES-256-GCM, HMAC-SHA-256, SHA-256,
 * SecureRandom, base64 — comes from the platform.
 *
 * The BouncyCastle **JCE provider is deliberately NOT installed**. Registering
 * `BouncyCastleProvider` would change which implementation every other consumer
 * in the process gets for every algorithm, which is a large, invisible blast
 * radius in exchange for nothing: the two primitives above are reached through
 * their lightweight classes directly.
 *
 * ## What is reimplemented, and why that is safe
 *
 * libsodium's `crypto_kx_*_session_keys` is not a primitive; it is a named
 * composition of X25519 and BLAKE2b-512, and this file reproduces that
 * composition exactly ([deriveSession]). That is the one place a hand-rolled
 * step could silently differ from the browser and the Apple clients, so it is
 * pinned to the shared golden vectors — both directions' keys, the SAS, the
 * commitment and the resume MAC — before anything opens a socket. No primitive
 * is invented anywhere in this file.
 */
object Crypto {

    // ── sizes ───────────────────────────────────────────────────────────────

    const val PUBLIC_KEY_BYTES = 32
    const val SECRET_KEY_BYTES = 32
    const val SESSION_KEY_BYTES = 32
    const val COMMIT_BYTES = 32
    const val COMMIT_NONCE_BYTES = 32
    const val AEAD_NONCE_BYTES = 12
    const val AEAD_TAG_BYTES = 16
    const val AEAD_TAG_BITS = AEAD_TAG_BYTES * 8

    /**
     * Domain separation for the message lane's keys.
     *
     * **17 bytes**, including the trailing NUL: `relayium-text-v1` is sixteen
     * ASCII characters plus U+0000. Two protocol documents said 18 until
     * 2026-09-07; the literal never changed, and a port that pads to 18 derives
     * a different key and decrypts nothing the browser sends.
     */
    val TEXT_KEY_DOMAIN: ByteArray = "relayium-text-v1\u0000".toByteArray(Charsets.US_ASCII)

    /** 24 bytes including the trailing NUL. */
    val RESUME_AUTH_DOMAIN: ByteArray = "relayium-resume-auth-v1\u0000".toByteArray(Charsets.US_ASCII)

    // `relayium-preupload-v1\0` (22 bytes) is deliberately absent. This client
    // does not announce `preupload/1`, so it can never legally receive frame
    // kind 12, and carrying the derivation for a capability it does not claim
    // would be code with no reachable caller and a key with no owner.

    init {
        // Sizes asserted at load rather than trusted: every one of them is a
        // constant of a wire two other implementations already speak, and a
        // silent disagreement here is a link that opens and then desynchronises.
        require(TEXT_KEY_DOMAIN.size == 17) { "text key domain must be 17 bytes" }
        require(RESUME_AUTH_DOMAIN.size == 24) { "resume-auth domain must be 24 bytes" }
    }

    // ── key pairs ───────────────────────────────────────────────────────────

    class KeyPair(val publicKey: ByteArray, private val secretKey: ByteArray) {
        internal fun secret(): ByteArray = secretKey
        /** Best effort: bounds how long the scalar sits in a heap dump. */
        fun destroy() = Bytes.wipe(secretKey)
    }

    private val random = SecureRandom()

    /**
     * A fresh `crypto_kx` keypair.
     *
     * libsodium stores the secret unclamped and clamps during the
     * multiplication; RFC 7748's `decodeScalar25519` inside BouncyCastle does
     * the same, so 32 random bytes plus `scalarMultBase` is byte-identical to
     * `crypto_kx_keypair`.
     *
     * A fresh pair per ESTABLISHMENT, never per app run: `relayium-link-v1.md`
     * section 5.5 — one link owns one set of session keys for its lifetime, and
     * a second link must not be able to reuse the first's nonce space.
     */
    fun generateKeyPair(): KeyPair {
        val secret = ByteArray(SECRET_KEY_BYTES)
        random.nextBytes(secret)
        val public = ByteArray(PUBLIC_KEY_BYTES)
        X25519.scalarMultBase(secret, 0, public, 0)
        return KeyPair(public, secret)
    }

    // ── session keys ────────────────────────────────────────────────────────

    enum class Role { INITIATOR, RESPONDER }

    /**
     * Everything one authenticated link is keyed with. Constructed once per
     * establishment and never rebuilt for a new batch or a reopened
     * conversation.
     */
    class SessionKeys internal constructor(
        internal val sendKey: ByteArray,
        internal val recvKey: ByteArray,
        internal val resumeAuthKey: ByteArray,
        internal val textSendKey: ByteArray,
        internal val textRecvKey: ByteArray,
    ) {
        internal val send = SecretKeySpec(sendKey, "AES")
        internal val recv = SecretKeySpec(recvKey, "AES")
        internal val textSend = SecretKeySpec(textSendKey, "AES")
        internal val textRecv = SecretKeySpec(textRecvKey, "AES")

        fun destroy() {
            Bytes.wipe(sendKey); Bytes.wipe(recvKey); Bytes.wipe(resumeAuthKey)
            Bytes.wipe(textSendKey); Bytes.wipe(textRecvKey)
        }
    }

    /** The peer offered a public key this protocol refuses to agree on. */
    class LowOrderKeyException : GeneralSecurityException(
        "relayium: peer public key produced an all-zero shared secret",
    )

    /**
     * libsodium `crypto_kx_client_session_keys` / `crypto_kx_server_session_keys`,
     * reproduced exactly.
     *
     * ```
     * q  = X25519(selfSecret, peerPublic)                       // refuse all-zero
     * h  = BLAKE2b-512(q || clientPublic || serverPublic)       // 64 bytes, unkeyed
     * initiator (client):  rx = h[0..32),  tx = h[32..64)
     * responder (server):  rx = h[32..64), tx = h[0..32)
     * ```
     *
     * Two details a port gets wrong and the golden vectors catch:
     *
     * 1. **The hash input is always `client_pk || server_pk`, in that order**,
     *    for BOTH roles. It is not "self then peer".
     * 2. **The halves are swapped by role, not the input.** That mirroring is
     *    what makes one side's `tx` the other's `rx` with no extra round trip.
     *
     * The all-zero check is not optional. `X25519.calculateAgreement` returns
     * false exactly when the peer offered a small-order point, which is
     * libsodium's own `crypto_scalarmult` contract; continuing would give both
     * ends a shared secret an attacker also knows.
     */
    @Throws(LowOrderKeyException::class)
    fun deriveSession(role: Role, self: KeyPair, peerPublic: ByteArray): SessionKeys {
        require(peerPublic.size == PUBLIC_KEY_BYTES) {
            "peer public key must be $PUBLIC_KEY_BYTES bytes, got ${peerPublic.size}"
        }
        val shared = ByteArray(PUBLIC_KEY_BYTES)
        val ok = X25519.calculateAgreement(self.secret(), 0, peerPublic, 0, shared, 0)
        if (!ok) {
            Bytes.wipe(shared)
            throw LowOrderKeyException()
        }
        val clientPublic = if (role == Role.INITIATOR) self.publicKey else peerPublic
        val serverPublic = if (role == Role.INITIATOR) peerPublic else self.publicKey

        val digest = Blake2bDigest(512)
        digest.update(shared, 0, shared.size)
        digest.update(clientPublic, 0, clientPublic.size)
        digest.update(serverPublic, 0, serverPublic.size)
        val both = ByteArray(2 * SESSION_KEY_BYTES)
        digest.doFinal(both, 0)
        Bytes.wipe(shared)

        val first = both.copyOfRange(0, SESSION_KEY_BYTES)
        val second = both.copyOfRange(SESSION_KEY_BYTES, 2 * SESSION_KEY_BYTES)
        Bytes.wipe(both)

        val rx = if (role == Role.INITIATOR) first else second
        val tx = if (role == Role.INITIATOR) second else first

        return SessionKeys(
            sendKey = tx.copyOf(),
            recvKey = rx.copyOf(),
            resumeAuthKey = resumeAuthKey(tx, rx),
            // NOT sorted, and it must not be: crypto_kx already mirrors tx/rx,
            // so hashing each locally lines the two directions up. Sorting would
            // collapse both onto one key and put two producers on one nonce
            // counter, which is the exact hazard the subkey removes.
            textSendKey = subkey(TEXT_KEY_DOMAIN, tx),
            textRecvKey = subkey(TEXT_KEY_DOMAIN, rx),
        ).also {
            Bytes.wipe(first); Bytes.wipe(second)
        }
    }

    /** `BLAKE2b-256(domain || sessionKey)`. */
    private fun subkey(domain: ByteArray, sessionKey: ByteArray): ByteArray {
        val digest = Blake2bDigest(256)
        digest.update(domain, 0, domain.size)
        digest.update(sessionKey, 0, sessionKey.size)
        val out = ByteArray(SESSION_KEY_BYTES)
        digest.doFinal(out, 0)
        return out
    }

    /**
     * The HMAC key both sides derive to the same value.
     *
     * SORTED, unlike the text subkeys, and for the opposite reason: this key is
     * SHARED, so it has to be symmetric. crypto_kx hands the peers mirrored
     * secrets, so hashing them in local order would give the two ends two
     * different keys; the pair as a SET is identical on both sides.
     */
    private fun resumeAuthKey(tx: ByteArray, rx: ByteArray): ByteArray {
        val first: ByteArray
        val second: ByteArray
        if (Bytes.compare(tx, rx) <= 0) { first = tx; second = rx } else { first = rx; second = tx }
        val digest = Blake2bDigest(256)
        digest.update(RESUME_AUTH_DOMAIN, 0, RESUME_AUTH_DOMAIN.size)
        digest.update(first, 0, first.size)
        digest.update(second, 0, second.size)
        val out = ByteArray(SESSION_KEY_BYTES)
        digest.doFinal(out, 0)
        return out
    }

    // ── AEAD ────────────────────────────────────────────────────────────────

    /**
     * The 12-byte nonce: four zero bytes, then a 64-bit big-endian counter.
     *
     * The counter is the frame's sequence number, so nonce uniqueness under a
     * key is exactly the sequence discipline the lanes enforce — never rewound,
     * never reused, and a resume may only skip FORWARD.
     */
    fun nonceFromSeq(seq: Long): ByteArray {
        require(seq >= 0) { "sequence numbers are non-negative; got $seq" }
        val nonce = ByteArray(AEAD_NONCE_BYTES)
        Bytes.writeUInt32(nonce, 4, (seq ushr 32) and 0xffff_ffffL)
        Bytes.writeUInt32(nonce, 8, seq and 0xffff_ffffL)
        return nonce
    }

    private fun aead(mode: Int, key: SecretKeySpec, seq: Long): Cipher =
        Cipher.getInstance("AES/GCM/NoPadding").apply {
            init(mode, key, GCMParameterSpec(AEAD_TAG_BITS, nonceFromSeq(seq)))
        }

    /** AES-256-GCM, combined mode: the 16-byte tag is appended. No AAD. */
    internal fun seal(key: SecretKeySpec, seq: Long, plaintext: ByteArray): ByteArray =
        aead(Cipher.ENCRYPT_MODE, key, seq).doFinal(plaintext)

    /** The inverse. An authentication failure throws; it is never a soft result. */
    @Throws(GeneralSecurityException::class)
    internal fun open(key: SecretKeySpec, seq: Long, ciphertext: ByteArray): ByteArray =
        aead(Cipher.DECRYPT_MODE, key, seq).doFinal(ciphertext)

    fun sealFile(keys: SessionKeys, seq: Long, plaintext: ByteArray): ByteArray =
        seal(keys.send, seq, plaintext)

    fun openFile(keys: SessionKeys, seq: Long, ciphertext: ByteArray): ByteArray =
        open(keys.recv, seq, ciphertext)

    fun sealText(keys: SessionKeys, seq: Long, plaintext: ByteArray): ByteArray =
        seal(keys.textSend, seq, plaintext)

    fun openText(keys: SessionKeys, seq: Long, ciphertext: ByteArray): ByteArray =
        open(keys.textRecv, seq, ciphertext)

    // ── commitment ──────────────────────────────────────────────────────────

    fun randomCommitNonce(): ByteArray = ByteArray(COMMIT_NONCE_BYTES).also(random::nextBytes)

    /** `BLAKE2b-256(publicKey || nonce)`. */
    fun commitKey(publicKey: ByteArray, nonce: ByteArray): ByteArray {
        val digest = Blake2bDigest(COMMIT_BYTES * 8)
        digest.update(publicKey, 0, publicKey.size)
        digest.update(nonce, 0, nonce.size)
        val out = ByteArray(COMMIT_BYTES)
        digest.doFinal(out, 0)
        return out
    }

    /**
     * Constant-time check that `commit` opens to `(publicKey, nonce)`.
     *
     * False on ANY mismatch, including a short or long commitment. A caller that
     * has no recorded commitment must treat that as a failure too — that check
     * belongs to the handshake state machine, which is where "no commit was ever
     * recorded" is distinguishable from "a commit was recorded and disagreed".
     */
    fun verifyCommit(commit: ByteArray, publicKey: ByteArray, nonce: ByteArray): Boolean {
        if (commit.size != COMMIT_BYTES) return false
        if (publicKey.size != PUBLIC_KEY_BYTES || nonce.size != COMMIT_NONCE_BYTES) return false
        return Bytes.constantTimeEquals(commit, commitKey(publicKey, nonce))
    }

    // ── SAS ─────────────────────────────────────────────────────────────────

    /**
     * The six-digit Short Authentication String.
     *
     * Sort the two raw public keys ascending, `BLAKE2b(8, a || b)`, XOR the two
     * big-endian uint32 halves, take modulo 1 000 000, zero-pad to six.
     *
     * The XOR is UNSIGNED. Reading either half into a signed `Int` and XORing
     * gives a negative value whose Kotlin `%` is also negative, and the code
     * comes out wrong exactly half the time — which a single hand-tested pairing
     * would very likely miss. `crypto-vectors.json` pins `705955`.
     *
     * Six digits, and NOT the pairing code, which is also six digits and
     * unrelated. `relayium-handshake-v1.md` says user-facing copy must never
     * conflate them; `PairCode` and this function deliberately return different
     * types so a UI cannot render one where it meant the other.
     */
    fun sas(selfPublic: ByteArray, peerPublic: ByteArray): String {
        val first: ByteArray
        val second: ByteArray
        if (Bytes.compare(selfPublic, peerPublic) <= 0) {
            first = selfPublic; second = peerPublic
        } else {
            first = peerPublic; second = selfPublic
        }
        val digest = Blake2bDigest(64)
        digest.update(first, 0, first.size)
        digest.update(second, 0, second.size)
        val out = ByteArray(8)
        digest.doFinal(out, 0)
        val hi = Bytes.readUInt32(out, 0)
        val lo = Bytes.readUInt32(out, 4)
        val value = (hi xor lo) % 1_000_000L
        return value.toString().padStart(6, '0')
    }

    // ── signalling tags ─────────────────────────────────────────────────────

    /**
     * `base64(HMAC-SHA-256(resumeAuth, utf8(payload)))`.
     *
     * Standard base64 with padding, so the result is exactly
     * [LinkProtocol.AUTH_TAG_LENGTH] characters — the length a verifier checks
     * before spending any work.
     */
    fun signAuth(keys: SessionKeys, payload: String): String {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(keys.resumeAuthKey, "HmacSHA256"))
        return Bytes.base64(mac.doFinal(payload.toByteArray(Charsets.UTF_8)))
    }

    /**
     * Verify a tag. Absent, malformed or wrong-length is **false**, never a
     * throw and never a pass — an "unauthenticated is acceptable" branch would
     * make the whole binding optional at the attacker's choosing.
     */
    fun verifyAuth(keys: SessionKeys, payload: String, tag: String?): Boolean {
        if (tag == null || tag.length != LinkProtocol.AUTH_TAG_LENGTH) return false
        val provided = Bytes.unbase64OrNull(tag) ?: return false
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(keys.resumeAuthKey, "HmacSHA256"))
        return Bytes.constantTimeEquals(provided, mac.doFinal(payload.toByteArray(Charsets.UTF_8)))
    }

    // ── integrity chain ─────────────────────────────────────────────────────

    /**
     * `h = SHA-256(h || chunk)`, starting from 32 zero bytes, over CHUNK_SIZE
     * logical chunks — regardless of how many DataChannel messages carried each
     * one.
     *
     * This is what `DONE_ENC` carries. It is deliberately NOT a plain SHA-256 of
     * the file: `relayium-link-v1.md` section 9.7.
     */
    fun chainStart(): ByteArray = ByteArray(32)

    fun chainAdvance(previous: ByteArray, chunk: ByteArray): ByteArray {
        val sha = MessageDigest.getInstance("SHA-256")
        sha.update(previous)
        sha.update(chunk)
        return sha.digest()
    }
}
